File: backend/app/utils/validation.py
import re

PASSWORD_MIN_LENGTH = 8


def validate_password_strength(password: str) -> str:
    """Raises ValueError with a human-readable message if the password is too weak."""
    if len(password) < PASSWORD_MIN_LENGTH:
        raise ValueError(f"Password must be at least {PASSWORD_MIN_LENGTH} characters long.")
    if not re.search(r"[a-z]", password):
        raise ValueError("Password must contain at least one lowercase letter.")
    if not re.search(r"[A-Z]", password):
        raise ValueError("Password must contain at least one uppercase letter.")
    if not re.search(r"\d", password):
        raise ValueError("Password must contain at least one digit.")
    if not re.search(r"[^\w\s]", password):
        raise ValueError("Password must contain at least one special character.")
    return password

File: frontend/features/auth/validation.ts
// Mirrors the backend's password policy (app/utils/validation.py) so users get
// immediate feedback; the backend re-validates regardless, as the source of truth.
export function getPasswordStrengthError(password: string): string | null {
  if (password.length < 8) return "Password must be at least 8 characters long.";
  if (!/[a-z]/.test(password)) return "Password must contain at least one lowercase letter.";
  if (!/[A-Z]/.test(password)) return "Password must contain at least one uppercase letter.";
  if (!/\d/.test(password)) return "Password must contain at least one digit.";
  if (!/[^\w\s]/.test(password)) return "Password must contain at least one special character.";
  return null;
}
